src/sparse/impl/KokkosSparse_spmv_impl.hpp

```diff
@@ -502,63 +508,606 @@ struct SPMV_MV_Transpose_Functor {
+      x1 += input.x1;
+    }
+
+    KOKKOS_INLINE_FUNCTION
+    value_type get(const ordinal_type vecIdx) {
+      switch (vecIdx) {
+      case 0:
+        return x0;
+      case 1:
+        return x1;
+      }
+
+      return 0;
+    }
+  };
+
+  SPMV_MV_Functor (const value_type alpha_,
+                   const AMatrix m_A_,
+                   const XVector m_x_,
+                   const value_type beta_,
+                   const YVector m_y_,
+                   const ordinal_type rows_per_team_) :
+    alpha (alpha_), m_A (m_A_), m_x (m_x_),
+    beta (beta_), m_y (m_y_),
+    rows_per_team (rows_per_team_),
+    numVecs (m_x_.extent(1)),
+    vecRemainder (m_x_.extent(1) % 16)
+  {
+    static_assert (static_cast<int> (XVector::rank) == 2,
+                   "XVector must be a rank 2 View.");
+    static_assert (static_cast<int> (YVector::rank) == 2,
+                   "YVector must be a rank 2 View.");
+  }
+
+  KOKKOS_INLINE_FUNCTION
+  void operator() (const team_member& dev) const
+  {
+    Kokkos::parallel_for(Kokkos::TeamThreadRange(dev,0,rows_per_team), [&] (const ordinal_type& loop) {
+        const ordinal_type iRow = static_cast<ordinal_type>(dev.league_rank()) * rows_per_team + loop;
+        if (iRow >= m_A.numRows ()) {
+          return;
+        }
+        const KokkosSparse::SparseRowViewConst<AMatrix> row = m_A.rowConst(iRow);
+        const ordinal_type row_length = static_cast<ordinal_type> (row.length);
+        ordinal_type remainder = vecRemainder;
+
+	for(ordinal_type vecOffset = 0; vecOffset < numVecs - 15; vecOffset += 16) {
+	  value16 sum = {};
+	  Kokkos::parallel_reduce(Kokkos::ThreadVectorRange(dev, row_length),
+				  [&] (const ordinal_type& entryIdx, value16& lsum) {
+
+				    // Extract data from current row
+				    const value_type val = alpha*(conjugate ? ATV::conj(row.value(entryIdx)) : row.value(entryIdx));
+				    const ordinal_type colIdx = row.colidx(entryIdx);
+
+				    // Perform dot product accros lhs columns
+				    lsum.x0  += m_x(colIdx, vecOffset +  0)*val;
+				    lsum.x1  += m_x(colIdx, vecOffset +  1)*val;
+				    lsum.x2  += m_x(colIdx, vecOffset +  2)*val;
+				    lsum.x3  += m_x(colIdx, vecOffset +  3)*val;
+				    lsum.x4  += m_x(colIdx, vecOffset +  4)*val;
+				    lsum.x5  += m_x(colIdx, vecOffset +  5)*val;
+				    lsum.x6  += m_x(colIdx, vecOffset +  6)*val;
+				    lsum.x7  += m_x(colIdx, vecOffset +  7)*val;
+				    lsum.x8  += m_x(colIdx, vecOffset +  8)*val;
+				    lsum.x9  += m_x(colIdx, vecOffset +  9)*val;
+				    lsum.x10 += m_x(colIdx, vecOffset + 10)*val;
+				    lsum.x11 += m_x(colIdx, vecOffset + 11)*val;
+				    lsum.x12 += m_x(colIdx, vecOffset + 12)*val;
+				    lsum.x13 += m_x(colIdx, vecOffset + 13)*val;
+				    lsum.x14 += m_x(colIdx, vecOffset + 14)*val;
+				    lsum.x15 += m_x(colIdx, vecOffset + 15)*val;
+				  }, sum);
+
+          Kokkos::parallel_for(Kokkos::ThreadVectorRange(dev, 16), [&] (const ordinal_type& vecIdx) {
```

Comment:
`ordinal_type` is probably `int`, so it's better to pass by value than by reference.